File: Solutions/Valid_Parenthesis_String/Valid_Parenthesis_String.js
var checkValidString = function (s) {
  const openStack = [];
  const starStack = [];

  for (let i = 0; i < s.length; i++) {
    if (s[i] === "(") {
      openStack.push(i);
    } else if (s[i] === "*") {
      starStack.push(i);
    } else {
      if (openStack.length > 0) {
        openStack.pop();
      } else if (starStack.length > 0) {
        starStack.pop();
      } else {
        return false;
      }
    }
  }

  let i = openStack.length - 1;
  let j = starStack.length - 1;

  while (openStack[i] < starStack[j]) {
    openStack.pop();
    starStack.pop();
    i--;
    j--;
  }

  if (openStack.length === 0) {
    return true;
  } else {
    return false;
  }
};

//Fastest
/**
 * @param {string} s
 * @return {boolean}
 */
var checkValidString = function (s) {
  var a = s.split("");
  return checkOneWay(a) && checkOneWay(invert(a));
};

var checkOneWay = function (a) {
  var lvl = 0;
  var stars = 0;
  for (var i = 0; i < a.length; i++) {
    var c = a[i];
    if (c === "(") {
      lvl++;
    }
    if (c === ")") {
      lvl--;
    }
    if (c === "*") {
      stars++;
    }
    if (lvl < 0) {
      stars--;
      lvl++;
    }
    if (stars < 0) {
      return false;
    }
  }
  return true;
};

var invert = function (a) {
  return a.reverse().map((c) => {
    if (c === "(") {
      return ")";
    } else if (c === ")") {
      return "(";
    } else {
      return c;
    }
  });
};

File: Solutions/Valid_Parenthesis_String/Valid_Parenthesis_String.cs
public class Solution {
    // Time Complexity: O(N+N+N+N)
    // Space Complexity: O(2N)
    public bool CheckValidString(string s) {
        int n = s.Length; 
        // need to know the location of the stars
        List<int> starIndicies = new List<int>();
		
		// using Tuple here, can be converted to a small class
        Stack<(char,int)> st = new Stack<(char, int)>(); 

        // O(N)
        for(int i = 0; i < n; i++){
            char c = s[i];
            if(c == '*')
                starIndicies.Add(i);
            else if(c == '(')
                st.Push((c, i));
            else if(c == ')' && st.Count() > 0 && st.Peek().Item1 == '(')
                st.Pop();
            else
                st.Push((c, i));
        }

        // O(N)
        // let's deal with what's left in the Stack
        while(st.Count > 0){
            if(starIndicies.Count() < 1)
                return false;

            var tup = st.Pop(); 
            if(tup.Item1 == '('){
                // find a star after index i, remove the star
                if(!FindNextStar(tup.Item2, starIndicies)){
                    return false; 
                }
            }
            else{
                // find a star after index i, remove the star
                if(!FindPrevStar(tup.Item2, starIndicies))
                    return false; 
            }
        }

        return true; 
    }

    // O(N)
    public bool FindNextStar(int index, IList<int> stars){
        foreach(int i in stars){
            if(i > index){
                stars.Remove(i);
                return true; 
            }
        }
        return false; 
    }

    // O(N)
    public bool FindPrevStar(int index, IList<int> stars){
        foreach(int i in stars){
            if(i < index){
                stars.Remove(i);
                return true; 
            }
        }
        return false; 
    }
}



//Fastest
public class Solution {
    public bool CheckValidString(string s) {
        //=======================
        int maxSum = 0 ;
        int minSum = 0 ;
        
        for(int i = 0 ; i < s.Length ;i++){
            if(s[i] == '('){
                minSum++;
            }
            else{
                minSum--;
            }
            if(s[i] == ')'){
                maxSum--;
            }
            else{
                maxSum++;
            }

            if(maxSum < 0 ){
                return false;
            }
            if(minSum < 0 ){
                minSum = 0;
            }
        }
        return minSum == 0;
    }
}